src/Spectrometer-OceanOptics_OceanDirect/main.py

```diff
@@ -0,0 +1,325 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+# SweepMe! driver
+# * Module: Spectrometer
+# * Instrument: Ocean Optics USB spectrometers
+
+
+import os
+import numpy as np
+from pysweepme.ErrorMessage import error
+from pysweepme.FolderManager import addFolderToPATH
+
+od_api_imported = False
+od_path_variable = "OCEANDIRECT_HOME"
+api_path_defined = od_path_variable in os.environ
+if api_path_defined:
+    od_path = os.environ[od_path_variable] + os.sep + "Python"
+    # Adding the OceanDirect Python API module to the environment so it can be imported
+    addFolderToPATH(od_path)
+
+    try:
+        from oceandirect import OceanDirectAPI
+        od_api_imported = True
+    except ImportError:
+        error()
+
+from pysweepme.EmptyDeviceClass import EmptyDevice
+
+
+class Device(EmptyDevice):
+
+    def __init__(self):
+
+        super().__init__()
+        
+        self.shortname = "OceanDirect"
+        
+        self.variables = ["Wavelength", "Intensity", "Integration time"]
+        self.plottype = [True, True, True]
+        self.savetype = [True, True, True]
+
+        self.calibrationfolder = self.get_folder("CALIBRATIONS")
+
+        self.trigger_modes = {
+            "Normal": 0,
+            "External Software Trigger": 1,
+            "External Synchronization Trigger": 2,
+            "Hardware Trigger": 3,
+            "Single-Shot Trigger": 4,
+        }
+
+        self.unique_ocean_direct_api_identifier = "OceanOptics_OceanDirectAPI"
+
+    def find_ports(self):
+
+        self.check_oceandirect_api()
+
+        # get OceanDirectAPI instance that is shared across multiple driver instances (independent of the model)
+        self.od = self.restore_parameter(self.unique_ocean_direct_api_identifier)
+        if self.od is None:
+            self.od = OceanDirectAPI.OceanDirectAPI()
+            self.store_parameter(self.unique_ocean_direct_api_identifier, self.od)
+
+        number_devices = self.od.find_usb_devices()
+        # print("Found devices:", number_devices)
+
+        dev_ids = self.od.get_device_ids()
+
+        # print("Device IDs:", dev_ids)
+
+        ports = []
+       
+        for dev_id in dev_ids:
+
+            spectrometer = self.od.open_device(dev_id)
+            serial_number = spectrometer.get_serial_number()
+            model = spectrometer.get_model()
+
+            port_serial_number = model + " - " + serial_number
+            ports.append(port_serial_number)
+
+            self.unique_spectrometer_identifier = "OceanDirectAPI_spectrometer_" + port_serial_number
+            self.store_parameter(self.unique_spectrometer_identifier, spectrometer)
+
+            # spectrometer.close_device()
+            # od.close_device(dev_id)
+
+        # print("Ports:", ports)
+
+        if len(ports) == 0:
+            self.message_Box(
+                "No spectrometer found! Please make sure it is connected and the OceanDirect API is installed.")
```

Comment:
Is this error message accurate?
When the OceanDirect API is not installed, the check_oceandirect_api function should already raise an exception above.

Reply:
Thanks for spotting it, changed it!